### ISSUE TITLE
Document work-around for move closures and mutable bindings...

### ISSUE THREAD

**Issue by GordonBGood** [rust-lang/rust#37383]
In the Book Section 4.23 Closures, the "move closures" portion demonstrates clearly how mutability is not transferred back to the captured mutable variable.  However, nowhere easily accessible is it explained that there is a work around for this by making the binding a `Rc<RefCell<T>>` instead of a `mut T`, thereby giving it internal mutability instead of external (a requirement of using Rc which contents would otherwise be treated as immutable) and thus allowing both the external binding and the binding captured by the closure to refer to and change the same data since that is what `Rc` does - wrap a reference to data in a reference-counted structure stored on the heap.

The following example if added would demonstrate this:

```
use std::rc::Rc;
use std::cell::RefCell;

let num = Rc::new(RefCell::new(5));

let num_copy = num.clone();
let add_num = move |x: i32| *num_copy.borrow_mut() += x;

add_num(5);

assert_eq!(10, *num.borrow());
```

It should be mentioned that `Cell` can be used instead of `RefCell` if the data type is `Copy`able as it is here for the primitive type and perhaps that also could be demonstrated using `get()` and `set()` instead of `borrow()`, `borrow_mut`, and the reference operator `*`

It should also be mentioned that for multi-threaded use `Arc` needs to be used instead of `Rc` and that the  data needs to be protected by `Mutex`'es or `RWLock`'s when necessary to prevent reading of partially updated data.

This allows one to implement algorithms requiring recursive variables where the current or the future (for use in thunks) values of the data are used in further calculations.

I spent quite a few hours in research before coming up with this trick, but now realize that others already know how to do this; I think it should be generally immediately known.

**Comment by steveklabnik** [comment]
I'm a bit torn, to be honest; the book isn't supposed to be an end-all, be-all list of ways to write stuff in Rust. This kinda feels like a "Rust patterns" book or something.

That said, I'm not totally opposed, either, so let's keep this open. PRs welcome!

**Comment by GordonBGood** [comment]
> the book isn't supposed to be an end-all, be-all list of ways to write stuff in Rust. This kinda feels like a "Rust patterns" book or something.

I see your point on the Book not being supposed to be a "Rust patterns" (we have "Rust by Example" for that), but OTOH it should be a reference where people in general, some new to the language, can find ways to get things done that may be easy in other languages.  I think this is one of those cases as Rust closures are quite different in use to those of other languages, most of which bind by reference (which is why they have a whole section in the Book).

Recursive variables have many uses such as the sharing fix point combinator in Haskell `fix f = let x = f x in x` where `x` is lazy, fix and f are functions, but as they require a binding capture, can't be done without using closures in Rust, and in order to be able to recurse on the state of the `T` variable need to use either `Rc<RefCell<T>>` or `Arc<Mutex<T>>` (or `Arc<RwLock<T>>`).

I see the current text saying that one can't use a move closure to do these things, perhaps leaving someone trying to do something like that with a negative feeling about Rust, whereas it is quite possible using the above wrappers.

In other words, it isn't really a pattern, it is a suggested work around to the limitation you have already pointed out.

**Comment by steveklabnik** [comment]
We are accepting small PRs for the first edition of the book, but not keeping issues open. https://github.com/rust-lang/book is where you'd send said PR into; please do so if you want to see this fixed. Thanks!
